### ISSUE TITLE
Matomo issues

### ISSUE THREAD

**Issue by egpbos** [NLeSC/guide#421]
There are currently two issues with our Matomo setup (introduced in #419):

1. We don't have an opt-out button (this is recommended by Matomo for GDPR compliance). I can't get the code for this to work in Docsify (it is included in the privacy page, but commented out for now).
2. More importantly: page visits are not properly logged. In particular, direct visits inside one browser session (tab/window/whatever) _are_ logged, but any subsequent page visits within that same session are not. This is probably because Docsify does something "smart", capturing the sub-page clicks, changing the address in the address bar and refreshing the page to contain the content you clicked on, but not actually doing any HTTP requests. The page just changes content through client-side Javascript. Apparently (and not surprisingly) this does not jive with Matomo's logging script.

We could investigate how to fix this inside Docsify, but we should not spend too much effort on this (e.g. writing a Docsify Matomo plugin would not be sustainable for us, but maybe there's an easy trick to make Matomo work with Docsify or Docsify's page update system could be slightly tweaked so that it emulates HTTP requests better...).

Also related to #405: with mkdocs I think we wouldn't have any of these problems, so that would also be a solution.

**Comment by egpbos** [comment]
@ewan-escience thoughts? Or @sverhoeven who I think set up the Docsify system originally.

**Comment by ewan-escience** [comment]
I suspect problem 2 can be solved by setting the [routerMode](https://docsify.js.org/#/configuration?id=routermode) setting to `history` (because every path is now after the pound sign in [https://guide.esciencecenter.nl/#/](https://guide.esciencecenter.nl/#/)). Note that this changes all paths and that all pages are then saved in the user's browser history.

However, when I try this out locally, deep linking does not work anymore, which makes sense, because there are no HTML pages at the paths.

**Comment by egpbos** [comment]
Unfortunately Matomo still doesn't log specific page visits. This is me visiting several pages and exiting to The Turing Way:

![Image](https://github.com/user-attachments/assets/8f3558de-a9ca-4d2c-81a1-4fb0b86e62d3)

It captures the exit correctly at least, but not the page URLs...

**Comment by egpbos** [comment]
Hey look at that, another tester has more luck it seems!

![Image](https://github.com/user-attachments/assets/d5cb05f0-badf-4b37-a866-f3ba2fa784dd)

Maybe it's something specific in my browser settings...

**Comment by sverhoeven** [comment]
Above is likely me from Utrecht.
Hmm, if disable ublock lite in Google Chrome 135 on Windows, I see it pinging `https://matomo.research.software/matomo.php?` with `https%3A%2F%2Fguide.esciencecenter.nl%2F%23%2Flanguage_guides%2Fjavascript` in search parameter. 


![Image](https://github.com/user-attachments/assets/4ceb8b3e-7cae-4e4f-ba0a-f0618300c8b7)

**Comment by egpbos** [comment]
Interesting, this is what happens when I click the H1 headers in the pages (which changes the URL to append `?id=c-and-c`):

![Image](https://github.com/user-attachments/assets/7594afe6-454f-4263-915d-a0cc79d74b7b)

**Comment by egpbos** [comment]
Ok, turning off Adblock Pro (Safari 18.3.1, MacOS 15.3.2) also changes things for the better. It still seems a bit odd: it only logs page visits after I visit the next page. So I start e.g. at Python, that does nothing in Matomo. Then I go to R, and then my Python visit shows up in Matomo...

**Comment by egpbos** [comment]
... ok not on main, because of branch protection 🙄

**Comment by egpbos** [comment]
Whoops 

![Image](https://github.com/user-attachments/assets/79e05a91-81f7-4840-aa48-0de20922cd5c)

**Comment by egpbos** [comment]
To be exact: the URL gets logged correctly now and at the right time (after clicking), but it gets logged together with the previous title! See this result:

![Image](https://github.com/user-attachments/assets/0592977f-f201-437a-b8b3-31fd8dc28946)

This log sequence was generated by first opening the main page directly via https://guide.esciencecenter.nl, then clicking Introduction in the menu and then clicking Best Practices in the menu.

**Comment by egpbos** [comment]
Could the issue be that the `document.title` variable is not yet updated at the time of the `hashchange` event?

**Comment by egpbos** [comment]
Indeed, my debugger says this is the case. The new title is not yet available at the event time... So how do we get it into Matomo? Note that the docs also don't use `document.title` in the event listener, but just a static `"My new title"`.

**Comment by egpbos** [comment]
Jeez, ok, while searching the Docsify source, I came across this fun undocumented piece of code 😄 https://github.com/docsifyjs/docsify/blob/7652bd471145a4f0af7b50cd377fa827eb0c4f20/src/plugins/matomo.js

I wonder if that does work properly though, since it uses the `hook.beforeEach` which I assume will give the same issue with the title... Going by [this plugin documentation] I would guess that `hook.done` would work better...

**Comment by egpbos** [comment]
The original problems in this issue are solved in #433. There are a few remaining cleanup issues:

1. We should remove the old (commented out) Matomo code which despite @sverhoeven's [comment](https://github.com/NLeSC/guide/pull/433#discussion_r2031122368) I forgot to remove.
2. Remove the logging of the netlify URL from Matomo config again.
3. We should make a PR with Docsify to modify the Matomo plugin there so we don't have to maintain it here.
4. After the above point, we should use the original Docsify Matomo plugin and remove our custom version.
5. @sverhoeven made a [good suggestion](https://github.com/NLeSC/guide/pull/433#discussion_r2031124894) to add a conditional that prevents downloading the Matomo script if you opted out of tracking (through the privacy page).
6. Somebody should check out why Chrome visits on macOS are not working (at least from my laptop... so probably it should be me checking that out).

**Comment by egpbos** [comment]
About 6.: "Do not track" was activated in my Chrome, deactivating it makes Matomo see me again, fixed!

**Comment by egpbos** [comment]
Closing this issue, follow-ups in #434 and #435.
